### PR TITLE
Support `save_hyperparameters()` in LightningModule dataclass

### LINKED ISSUE #3494: add support for save_hyperparameters with Python Data Class
Python [Data Classes](https://docs.python.org/3/library/dataclasses.html) are convenient in that they automatically generate a bunch of boilerplate code for assigning data to a class. They are particularly useful for PyTorch models that have a lot of hyperparameters and thus a lot of boilerplate.

## 🐛 Bug

I believe https://github.com/PyTorchLightning/pytorch-lightning/pull/1896 introduced a new bug: when using a data class, `save_hyperparameters` no longer works since it depends on init args and we instead use `__post_init__` with dataclasses. Explicitly passing strings does not work either. Perhaps when passing 

#### Code sample
```python
import pytorch_lightning as pl
from dataclasses import dataclass

@dataclass()
class ConvDecoder(pl.LightningModule):
    imageChannels:int = 3
        
    def __post_init__(self):
        super().__init__()
        # both fail
#         self.save_hyperparameters()
        self.save_hyperparameters('imageChannels')
        
model = ConvDecoder()
model.hparams
```

### Expected behavior

There should be a way to use `save_hyperparameters` with Data Classes

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #3494

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/parsing.py [hunk 1]
```diff
@@ -197,7 +198,11 @@ def save_hyperparameters(
 
     if not frame:
         frame = inspect.currentframe().f_back
-    init_args = get_init_args(frame)
+
+    if is_dataclass(obj):
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/parsing.py:202]: @awaelchli I don't think we are using this `save_hyperparameters` function, but the one in the Lightning module. I guess this was decoupled for an use case in Flash.

Comment (awaelchli) [reply pytorch_lightning/utilities/parsing.py:202]: in what sense do you mean we are not using it. the test that I added fails on master.

File: tests/models/test_hparams.py [hunk 2]
```diff
@@ -719,3 +721,21 @@ def test_empty_hparams_container(tmpdir):
     assert not model.hparams
     model = HparamsNamespaceContainerModel(Namespace())
     assert not model.hparams
+
+
+@dataclass
+class DataClassModel(BoringModel):
+
+    mandatory: int
+    optional: Optional[str] = "optional"
```
Comment (kaushikb11) [inline tests/models/test_hparams.py:730]:
```suggestion
    optional: str = "optional"
```

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/utilities/parsing.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/parsing.py:202]: Nvm, just checked, it was divided.

**Comment by kaushikb11** [comment]
Tests are failing with the latest version. Looking into it!

**Comment by awaelchli** [comment]
@kaushikb11 I think Thomas has the fix here in #7976 in the reinforce file, the Iterable typehint. You could try that.

**Comment by kaushikb11** [comment]
@awaelchli I remembered seeing it in PT 1.9 release notes.

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
